parsing/builtin_attributes.ml

```diff
@@ -15,6 +15,83 @@
 
 open Asttypes
 open Parsetree
+open Ast_helper
+
+
+module Attribute_table = Hashtbl.Make (struct
+  type t = string with_loc
+
+  let hash : t -> int = Hashtbl.hash
+  let equal : t -> t -> bool = (=)
+end)
+let unused_attrs = Attribute_table.create 128
+let mark_used t = Attribute_table.remove unused_attrs t
+
+(* [attr_order] is used to issue unused attribute warnings in the order the
+   attributes occur in the file rather than the random order of the hash table
+*)
+let attr_order a1 a2 =
+  match String.compare a1.loc.loc_start.pos_fname a2.loc.loc_start.pos_fname
+  with
+  | 0 -> Int.compare a1.loc.loc_start.pos_cnum a2.loc.loc_start.pos_cnum
+  | n -> n
+
+let warn_unused () =
+  let keys = List.of_seq (Attribute_table.to_seq_keys unused_attrs) in
+  let keys = List.sort attr_order keys in
+  List.iter (fun sloc ->
+    Location.prerr_warning sloc.loc (Warnings.Misplaced_attribute sloc.txt))
+    keys
+
+(* These are the attributes that are tracked in the builtin_attrs table for
+   misplaced attribute warnings. *)
+let builtin_attrs =
+  [ "alert"
+  ; "boxed"
+  ; "deprecated"
+  ; "deprecated_mutable"
+  ; "explicit_arity"
+  ; "immediate"
+  ; "immediate64"
+  ; "inline"
+  ; "inlined"
+  ; "noalloc"
+  ; "poll"
+  ; "ppwarning"
+  ; "specialise"
+  ; "specialised"
+  ; "tailcall"
+  ; "tail_mod_cons"
+  ; "unboxed"
+  ; "untagged"
+  ; "unrolled"
+  ; "warnerror"
+  ; "warning"
+  ; "warn_on_literal_pattern"
+  ]
+
+let builtin_attrs =
+  let tbl = Hashtbl.create 128 in
+  List.iter (fun attr -> Hashtbl.add tbl attr ()) builtin_attrs;
+  tbl
+
+let drop_ocaml_attr_prefix s =
+  let len = String.length s in
+  if String.starts_with ~prefix:"ocaml." s && len > 6 then
+    String.sub s 6 (len - 6)
+  else
+    s
```

Comment:
Note that it seems that in a handful of cases, we could avoid the allocation by having a `attr_equal` function rather than using the `String.equal (drop_ocaml_attr_prefix x) y` pattern.

Reply:
Good idea - I've made this change, leaving only the allocation in `is_builtin_attr`.  This seems like a fine compromise.